rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,681 @@
+migrations.](https://github.com/elastic/kibana/blob/75444a9f1879c5702f9f2b8ad4a70a3a0e75871d/docs/setup/upgrade/upgrade-migrations.asciidoc#handling-errors-during-saved-object-migrations)
+ - [Rolling back to a previous version of Kibana.](https://github.com/elastic/kibana/blob/75444a9f1879c5702f9f2b8ad4a70a3a0e75871d/docs/setup/upgrade/upgrade-migrations.asciidoc#rolling-back-to-a-previous-version-of-kib) 
+ - Server logs from failed migrations.
+
+**Assumptions and tradeoffs:**
+1. The simplicity of idempotent, coordination-free migrations outweighs the
+   restrictions this will impose on the kinds of migrations we're able to
+   support in the future. See (4.3.1)
+2. Maintaining the upgrade behaviour and rollback procedures of previous
+   Kibana 7.x releases is more important than introducing enhancements like:
+   1. (2.6) _mixed Kibana versions shouldn't cause data loss_.
+   2. In-place migrations to support > 10k saved objects, reduce the downtime
+      window, and slightly improve the reliability. See (4.5.1)
+3. For 8.x, in-place migrations to support > 10k saved objects, reduce the
+   downtime window, and slightly improve the reliability is more important
+   than (2.7) _maintain read-only functionality during the downtime
+   window_. See (4.5.1).
+4. A saved object type (and it's associated migrations) will only ever be
+   owned by one plugin. If pluginA registers saved object type `plugin_a_type`
+   then pluginB must never register that same type, even if pluginA is
+   disabled. Although we cannot enforce it on third-party plugins, breaking
+   this assumption may lead to data loss.
+
+## 4.1 Discover and remedy potential failures before any downtime
+
+> Achieves goals: (2.3)
+> Mitigates errors: (3.1), (3.2)
+
+1. Introduce a CLI option to perform a dry run migration to allow
+   administrators to locate and fix potential migration failures without
+   taking their existing Kibana node(s) offline.
+2. To have the highest chance of surfacing potential failures such as low disk
+   space, dry run migrations should not be mere simulations. A dry run should
+   perform a real migration in a way that doesn’t impact the existing Kibana
+   cluster.
+3. The CLI should generate a migration report to make it easy to create a
+   support request from a failed migration dry run.
+    1. The report would be an NDJSON export of all failed objects.
+    2. If support receives such a report, we could modify all the objects to
+       ensure the migration would pass and send this back to the client.
+    3. The client can then import the updated objects using the standard Saved
+       Objects NDJSON import and run another dry run to verify all problems
+       have been fixed.
+4. Make running dry run migrations a required step in the upgrade procedure
+   documentation.
+5. (Optional) Add dry run migrations to the standard cloud upgrade procedure?
+
+## 4.2 Tag objects as “invalid” if their transformation fails
```

Comment:
I've given this more thought and this really is a tradeoff between data consistency and availability. When a transform fails we cannot be sure if it was invalid data or a bug in the transform function so we cannot be sure that it is safe to simply ignore this object. Queries on the saved object type might suddenly exclude this object because it wasn't successfully migrated. The impact of this on users vary significantly between plugins so we cannot prefer availability for all plugin types.

It would be useful for plugins to be able to specify the behaviour they prefer, availability or consistency, but this also has problems. If there's a bug in a migration transform and e.g. all dashboards failed to migrate while this type preferred availability, Kibana would start up successfully and accept traffic. A user must now either delete and re-create all dashboards, wait for an emergency bug fix or rollback Kibana. But because Kibana is already serving traffic a rollback will loose data that's impossible to recover.

In such a scenario it would be preferable if the migration failed and the user was immediately able to rollback Kibana to a previous working version. This would mean a failed upgrade but wouldn't cause any downtime outside the scheduled maintenance window.